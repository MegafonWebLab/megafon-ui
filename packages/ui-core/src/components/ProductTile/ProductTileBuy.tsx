import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './style/ProductTileBuy.less';
import Button, { IButtonProps } from '../Button/Button';
import TextLink from '../TextLink/TextLink';
import { TLinkTargetType } from './ProductTile';

export interface IProductTileBuyProps {
    /** Class name */
    className?: string;
    /** Buy link */
    buyLink?: string;
    /** Buy Link target */
    buyLinkTarget?: TLinkTargetType;
    /** Buy button text */
    buyButtonText?: string;
    /** Show buy button */
    showBuyButton?: boolean;
    /** Connect link */
    connectLink?: string;
    /** Connect Link target */
    connectLinkTarget?: TLinkTargetType;
    /** Connect button text */
    connectButtonText?: string;
    /** Show connect button */
    showConnectButton?: boolean;
    /** button border */
    buttonBorder?: IButtonProps['border'];
    /** button font color */
    buttonFontColor?: IButtonProps['fontColor'];
    /** button background color */
    buttonPassiveColor?: IButtonProps['passiveColor'];
    /** Connect hander */
    onClickConnect?(e: React.SyntheticEvent<EventTarget>): void;
    /** Buy hander */
    onClickBuy?(e: React.SyntheticEvent<EventTarget>): void;
}

const cn = cnCreate('mfui-product-tile-buy');
class ProductTileBuy extends React.Component<IProductTileBuyProps> {
    static propTypes = {
        buyLink: PropTypes.string,
        buyLinkTarget: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
        buyButtonText: PropTypes.string,
        showBuyButton: PropTypes.bool,
        connectLink: PropTypes.string,
        connectLinkTarget: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
        connectButtonText: PropTypes.string,
        showConnectButton: PropTypes.bool,
        buttonBorder: Button.propTypes.border,
        buttonFontColor: Button.propTypes.fontColor,
        buttonPassiveColor: Button.propTypes.passiveColor,
        onClickBuy: PropTypes.func,
        onClickConnect: PropTypes.func,
    };

    static defaultProps: Partial<IProductTileBuyProps> = {
        buyButtonText: 'Купить',
        showBuyButton: true,
        connectButtonText: 'Перейти на тариф',
        showConnectButton: true,
        buyLinkTarget: '_self',
        connectLinkTarget: '_self',
    };

    render() {
        const {
            className,
            buyLink,
            buyButtonText,
            showBuyButton,
            connectLink,
            connectButtonText,
            showConnectButton,
            buttonBorder,
            buttonFontColor,
            buttonPassiveColor,
            onClickBuy,
            onClickConnect,
            buyLinkTarget,
            connectLinkTarget,
        } = this.props;

        return (
            <div className={cn('', {}, className)}>
                {showBuyButton &&
                    <Button
                        className={cn('button', { 'single-button': !showConnectButton })}
                        passiveColor={buttonPassiveColor}
                        hoverColor="green"
                        sizeAll="medium"
                        href={buyLink}
                        target={buyLinkTarget}
                        border={buttonBorder}
                        fontColor={buttonFontColor}
                        onClick={onClickBuy}
                    >
                        {buyButtonText}
                    </Button>
                }
                {showConnectButton &&
                    <TextLink
                        className={cn('detail-link')}
                        href={connectLink}
                        target={connectLinkTarget}
                        onClick={onClickConnect}
                    >
                        {connectButtonText}
                    </TextLink>
                }
            </div>
        );
    }
}

export default ProductTileBuy;
