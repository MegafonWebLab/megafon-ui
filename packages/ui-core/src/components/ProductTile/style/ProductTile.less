@import '~styles/base.less';

@b: e('.mfui-product-tile');

@{b} {
    display: flex;
    flex: 1 1 100%;
    flex-direction: column;
    justify-content: space-between;
    box-sizing: border-box;
    position: relative;
    border-radius: 10px;
    box-shadow: 0 1px 4px 0 rgba(0, 0, 0, 0.2);
    background-color: @clearWhite;
    padding: 32px 8px 16px;

    &:hover {
        box-shadow: 0 4px 10px 0 rgba(0, 0, 0, 0.2);
    }

    @media @mobileSM {
        &:not(:last-child) {
            margin-bottom: 24px;
        }
    }

    @media @mobileB {
        padding-top: 46px;
    }

    @media @desktopS {
        padding-top: 34px;
    }

    @media @desktopMU {
        padding: 34px 12px 20px;
    }

    &__content {
        margin-bottom: 25px;
    }

    &__constructor {
        background-color: lighten(@spbSky1, 0.5);
        padding: 24px 16px 16px;
        margin-top: 10px;
        position: relative;

        &:before {
            border-top-color: lighten(@spbSky1, 0.5);
            border-left-color: lighten(@spbSky1, 0.5);
            border-bottom-color: transparent;
            border-right-color: transparent;
            border-width: 5px;
            border-style: solid;
            content: '';
            left: 50%;
            position: absolute;
            top: -4px;
            transform: translateX(-50%) rotate(45deg);
        }

        @media @mobileB {
            padding: 16px 8px;
        }

        @media @desktopS {
            padding: 22px 8px;
        }

        @media @desktopMU {
            padding-left: 12px;
            padding-right: 12px;
        }
    }

    &__title {
        font-size: 20px;
        font-weight: 500;
        text-align: center;
        line-height: 28px;
        letter-spacing: 0.5px;
        min-height: 47px;

        @media @mobileSM {
            min-height: auto;
        }

        @media @mobileB {
            font-size: 18px;
            line-height: 24px;
        }

        @media @desktopS {
            font-size: 18px;
            line-height: 24px;
        }

        @media @desktopMU {
            min-height: 56px;
        }
    }

    &__link-wrapper {
        text-align: center;
        margin-bottom: 18px;

        @media @mobileSM {
            margin-bottom: 17px;
        }

        @media @desktopS {
            margin-bottom: 26px;
        }
    }

    &_active {
        box-shadow: inset 0 0 0 2px @greenActive, 0 4px 10px 0 rgba(0, 0, 0, 0.2);
    }

    &_active:hover {
        box-shadow: inset 0 0 0 2px @greenActive, 0 4px 10px 0 rgba(0, 0, 0, 0.2);
    }
}
