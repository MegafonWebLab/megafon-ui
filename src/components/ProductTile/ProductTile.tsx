import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './style/ProductTile.less';
import TextLink from '../TextLink/TextLink';
import Hint from './ProductTileHint';
import Buy from './ProductTileBuy';
import Cashback from './ProductTileCashback';
import Static from './ProductTileStatic';
import Dynamic from './ProductTileDynamic';
import Price from './ProductTilePrice';
import Options from './ProductTileOptions';

export interface IOption {
    title: string;
    caption?: string;
    value?: string;
    unit?: string;
    footnote?: string;
    svgIcon?: JSX.Element;
}

export interface IPayment {
    value: string;
    unitExtra: string;
    unitValue: string;
    discount?: string;
}

export interface IServicePack {
    buyLink: string;
    shopTag: string;
    calls: {
        value: number;
        unit: string;
    };
    traffic: {
        value: number;
        unit: string;
    };
    options: IOption[];
    payment: IServicePackPayment;
}

export interface IServicePackPayment {
    value: string;
    discount?: string;
}

export interface IPack {
    value: number;
    unit: string;
    title: string;
    isDelim: boolean;
}

export interface ICashback {
    title: string;
    value: number;
    unit: string;
}

export interface IFirstParam {
    items: IOption[];
}

export interface ISwitcher {
    calls: string[];
    traffic: string[];
}

interface IProductTileProps {
    /** Class name */
    className?: string;
    /** Tile */
    title: string;
    /** Top badge title */
    topBadgeTitle?: string;
    /** Second params head */
    secondParamsHead?: string;
    /** Shop tag */
    shopTag?: string;
    /** Start calls index */
    startCallsIndex?: number;
    /** Start traffic index */
    startTrafficIndex?: number;

    /** More link */
    link?: string;
    /** More link text */
    moreLinkText?: string;
    /** Show more link */
    showMoreLink?: boolean;

    /** Buy link */
    buyLink?: string;
    /** Buy button text */
    buyButtonText?: string;
    /** Show buy button */
    showBuyButton?: boolean;

    /** Connect link */
    connectLink?: string;
    /** Connect button text */
    connectButtonText?: string;
    /** Show connect button */
    showConnectButton?: boolean;

    /** Payment */
    payment: IPayment;
    /** Packs */
    packs?: IPack[];
    /** Second params */
    secondParams: IOption[];
    /** Cashback */
    cashback: ICashback;
    /** First params */
    firstParams: IFirstParam;
    /** Service packs */
    servicePacks?: Array<Partial<IServicePack>>;
    /** Info - object type - return with onClickConnect, onClickBuy */
    info: {};
    /** Connect handler */
    onClickConnect?(info: {}, e: React.SyntheticEvent<EventTarget>): void;
    /** Buy handler */
    onClickBuy?(info: {}, e: React.SyntheticEvent<EventTarget>): void;
    /** More handler */
    onClickMore?(info: {}, e: React.SyntheticEvent<EventTarget>): void;
    /** Bubble handler */
    onClickBubble?(info: {}): void;
}

interface IProductTileState {
    switcher: ISwitcher;
    currentPack: Partial<IServicePack>;
    callsValue: number;
    trafficValue: number;
    price: string;
    discount: string;
    options: IOption[];
    buyLink: string;
}

const cn = cnCreate('mfui-product-tile');
class ProductTile extends React.Component<IProductTileProps, IProductTileState> {
    static propTypes = {
        title: PropTypes.string.isRequired,
        topBadgeTitle: PropTypes.string,
        secondParamsHead: PropTypes.string,
        shopTag: PropTypes.string,
        startCallsIndex: PropTypes.number,
        startTrafficIndex: PropTypes.number,
        link: PropTypes.string,
        moreLinkText: PropTypes.string,
        showMoreLink: PropTypes.bool,
        buyLink: PropTypes.string,
        buyButtonText: PropTypes.string,
        showBuyButton: PropTypes.bool,
        connectLink: PropTypes.string,
        connectButtonText: PropTypes.string,
        showConnectButton: PropTypes.bool,
        payment: PropTypes.shape({
            value: PropTypes.string.isRequired,
            unitExtra: PropTypes.string,
            unitValue: PropTypes.string,
            discount: PropTypes.string,
        }).isRequired,
        packs: PropTypes.arrayOf(PropTypes.shape({
            value: PropTypes.number,
            unit: PropTypes.string,
        })),
        secondParams: PropTypes.arrayOf(PropTypes.shape({
            title: PropTypes.string,
            footnote: PropTypes.string,
            value: PropTypes.string,
            unit: PropTypes.string,
            svgIcon: PropTypes.element,
        })).isRequired,
        cashback: PropTypes.shape({
            title: PropTypes.string,
            value: PropTypes.number,
            unit: PropTypes.string,
        }).isRequired,
        firstParams: PropTypes.shape({
            items: PropTypes.arrayOf(PropTypes.shape({
                title: PropTypes.string,
                caption: PropTypes.string,
                svgIcon: PropTypes.element,
            })),
        }).isRequired,
        servicePacks: PropTypes.arrayOf(PropTypes.shape({
            buyLink: PropTypes.string,
            calls: PropTypes.shape({
                value: PropTypes.number,
                unit: PropTypes.string,
            }),
            traffic: PropTypes.shape({
                value: PropTypes.number,
                unit: PropTypes.string,
            }),
            options: PropTypes.arrayOf(PropTypes.shape({
                title: PropTypes.string,
                caption: PropTypes.stirng,
                value: PropTypes.string,
                unit: PropTypes.string,
                footnote: PropTypes.string,
                svgIcon: PropTypes.element,
            })),
            payment: PropTypes.shape({
                value: PropTypes.string.isRequired,
                discount: PropTypes.string,
            }),
        })),
        info: PropTypes.object,
        onClickConnect: PropTypes.func,
        onClickBuy: PropTypes.func,
        onClickMore: PropTypes.func,
        onClickBubble: PropTypes.func,
    };

    static defaultProps: Partial<IProductTileProps> = {
        moreLinkText: 'Подробнее',
        showMoreLink: true,
        servicePacks: [],
        info: {},
        startCallsIndex: 0,
        startTrafficIndex: 0,
    };

    constructor(props: IProductTileProps) {
        super(props);

        const switcher = this.getSwitcherValues();
        const defaultCallsValue = Number(switcher.calls[props.startCallsIndex!]);
        const defaultTrafficValue = Number(switcher.traffic[props.startCallsIndex!]);
        const currentPack = this.getCurrentPack(defaultCallsValue, defaultTrafficValue);
        const { payment, options, buyLink } = currentPack;

        this.state = {
            switcher,
            currentPack,
            callsValue: defaultCallsValue,
            trafficValue: defaultTrafficValue,
            price: payment ? payment.value : props.payment.value,
            discount: payment ? payment.discount || '' : props.payment.discount || '',
            options: options || props.secondParams,
            buyLink: buyLink || props.buyLink || '',
        };
    }

    handleClickConnect = (e: React.SyntheticEvent<EventTarget>) => {
        const { onClickConnect, info } = this.props;
        const { currentPack: { calls, traffic } } = this.state;
        const callsValue = calls ? calls.value : 0;
        const trafficValue = traffic ? traffic.value : 0;

        onClickConnect && onClickConnect({ ...info, callsValue, trafficValue }, e);
    }

    handleClickBuy = (e: React.SyntheticEvent<EventTarget>) => {
        const { onClickBuy, info, shopTag } = this.props;
        const { currentPack: { calls, traffic, shopTag: packShopTag } } = this.state;
        const callsValue = calls ? calls.value : 0;
        const trafficValue = traffic ? traffic.value : 0;
        const currentShopTag = packShopTag || shopTag || '';

        onClickBuy && onClickBuy({ ...info, callsValue, trafficValue, shopTag: currentShopTag }, e);
    }

    handleClickMore = (e: React.SyntheticEvent<EventTarget>) => {
        const { onClickMore, info } = this.props;

        onClickMore && onClickMore({ ...info }, e);
    }

    handleClickBubble = () => {
        const { onClickBubble, info } = this.props;

        onClickBubble && onClickBubble({ ...info });
    }

    handleChangeCalls = (_e: React.SyntheticEvent<EventTarget>, value: string): boolean => {
        const currentValue = Number(value);
        const { trafficValue } = this.state;
        const currentPack = this.getCurrentPack(currentValue, trafficValue);

        if (Object.keys(currentPack).length === 0) {
            return false;
        }

        this.setState({
            ...this.getRestState(currentPack),
            callsValue: currentValue,
        });

        return true;
    }

    handleChangeTraffic = (_e: React.SyntheticEvent<EventTarget>, value: string): boolean => {
        const currentValue = Number(value);

        const { callsValue } = this.state;
        const currentPack = this.getCurrentPack(callsValue, currentValue);

        if (Object.keys(currentPack).length === 0) {
            return false;
        }

        this.setState({
            ...this.getRestState(currentPack),
            trafficValue: currentValue,
        });

        return true;
    }

    getRestState(currentPack: Partial<IServicePack>): object {
        return {
            currentPack,
            price: currentPack.payment!.value,
            discount: currentPack.payment!.discount,
            options: currentPack.options!,
            buyLink: currentPack.buyLink!,
        };
    }

    getCurrentPack(callsValue: number, trafficValue: number): Partial<IServicePack> {
        const { servicePacks } = this.props;
        const [ currentPack = {} ] = servicePacks!.filter(pack => (
            pack.calls!.value === callsValue && pack.traffic!.value === trafficValue
        ));

        return currentPack;
    }

    getSwitcherValues() {
        const { servicePacks } = this.props;
        const calls = {};
        const traffic = {};

        servicePacks!.forEach(items => {
            calls[String(items.calls!.value)] = true;
            traffic[String(items.traffic!.value)] = true;
        });

        return {
            calls: Object.keys(calls).sort((a, b) => Number(a) - Number(b)),
            traffic: Object.keys(traffic).sort((a, b) => Number(a) - Number(b)),
        };
    }

    renderStatic() {
        const { packs } = this.props;

        if (!packs) {
            return null;
        }

        return <Static packs={packs} />;
    }

    renderDynamic() {
        const {
            switcher,
            currentPack,
        } = this.state;
        const {
            startCallsIndex,
            startTrafficIndex,
        } = this.props;

        return (
            <Dynamic
                startCallsIndex={startCallsIndex!}
                startTrafficIndex={startTrafficIndex!}
                currentPack={currentPack}
                switcher={switcher}
                onChangeCalls={this.handleChangeCalls}
                onChangeTraffic={this.handleChangeTraffic}
            />
        );
    }

    renderTitle() {
        const { title, link } = this.props;

        return (
            <h2 className={cn('title')}>
                <TextLink
                    className={cn('title-link')}
                    href={link}
                    underlineStyle="none"
                    color="black"
                    target="_blank"
                    onClick={this.handleClickMore}
                >
                    {title}
                </TextLink>
            </h2>
        );
    }

    renderLink() {
        const { link = '', showMoreLink, moreLinkText } = this.props;

        if (!showMoreLink || !moreLinkText) {
            return null;
        }

        const linkWithHash = this.formHashLink(link);

        return (
            <div className={cn('link-wrapper')}>
                <TextLink
                    className={cn('detail-link')}
                    href={linkWithHash}
                    target="_blank"
                    onClick={this.handleClickMore}
                >
                    {moreLinkText}
                </TextLink>
            </div>
        );
    }

    formHashLink(link: string) {
        const { currentPack: { shopTag } } = this.state;
        return !shopTag ? link : `${link}#${shopTag}`;
    }

    render() {
        const {
            servicePacks,
            cashback,
            firstParams,
            className,
            buyButtonText,
            connectButtonText,
            topBadgeTitle,
            secondParamsHead,
            showConnectButton,
            showBuyButton,
            connectLink,
            payment: { unitExtra, unitValue },
        } = this.props;
        const { price, discount, options, buyLink } = this.state;
        const isServicePacks = !!servicePacks!.length;

        return (
            <div className={cn('', { constructor: isServicePacks }, className)}>
                {isServicePacks && !!topBadgeTitle && <Hint title={topBadgeTitle}/>}
                <div className={cn('content')}>
                    {this.renderTitle()}
                    {this.renderLink()}
                    <Cashback {...cashback} />
                    <Price value={price} discount={discount} unitExtra={unitExtra} unitValue={unitValue} />
                    <div className={cn('constructor')}>
                        {isServicePacks ? this.renderDynamic() : this.renderStatic()}
                        <Options options={firstParams.items} />
                    </div>
                    <Options options={options} head={secondParamsHead} onClickBubble={this.handleClickBubble}/>
                </div>
                <Buy
                    buyLink={buyLink}
                    connectLink={connectLink}
                    buyButtonText={buyButtonText}
                    connectButtonText={connectButtonText}
                    showBuyButton={showBuyButton}
                    showConnectButton={showConnectButton}
                    onClickBuy={this.handleClickBuy}
                    onClickConnect={this.handleClickConnect}
                />
            </div>
        );
    }
}

export default ProductTile;
