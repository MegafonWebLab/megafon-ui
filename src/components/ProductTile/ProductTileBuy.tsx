import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './style/ProductTileBuy.less';
import Button, { IButtonProps } from '../Button/Button';
import TextLink from '../TextLink/TextLink';

export interface IProductTileBuyProps {
    /** Class name */
    className?: string;
    /** Buy link */
    buyLink?: string;
    /** Buy button text */
    buyButtonText?: string;
    /** Show buy button */
    showBuyButton?: boolean;
    /** Connect link */
    connectLink?: string;
    /** Connect button text */
    connectButtonText?: string;
    /** Show connect button */
    showConnectButton?: boolean;
    /** button border */
    buttonBorder?: IButtonProps['border'];
    /** button font color */
    buttonFontColor?: IButtonProps['fontColor'];
    /** button background color */
    buttonPassiveColor?: IButtonProps['passiveColor'];
    /** Connect hander */
    onClickConnect?(e: React.SyntheticEvent<EventTarget>): void;
    /** Buy hander */
    onClickBuy?(e: React.SyntheticEvent<EventTarget>): void;
}

const cn = cnCreate('mfui-product-tile-buy');
class ProductTileBuy extends React.Component<IProductTileBuyProps> {
    static propTypes = {
        buyLink: PropTypes.string,
        buyButtonText: PropTypes.string,
        showBuyButton: PropTypes.bool,
        connectLink: PropTypes.string,
        connectButtonText: PropTypes.string,
        showConnectButton: PropTypes.bool,
        buttonBorder: Button.propTypes.border,
        buttonFontColor: Button.propTypes.fontColor,
        buttonPassiveColor: Button.propTypes.passiveColor,
        onClickBuy: PropTypes.func,
        onClickConnect: PropTypes.func,
        onClickChoose: PropTypes.func,
    };

    static defaultProps: Partial<IProductTileBuyProps> = {
        buyButtonText: 'Купить',
        showBuyButton: true,
        connectButtonText: 'Перейти на тариф',
        showConnectButton: true,
    };

    render() {
        const {
            className,
            buyLink,
            buyButtonText,
            showBuyButton,
            connectLink,
            connectButtonText,
            showConnectButton,
            buttonBorder,
            buttonFontColor,
            buttonPassiveColor,
            onClickBuy,
            onClickConnect,
        } = this.props;

        return (
            <div className={cn('', {}, className)}>
                {showBuyButton &&
                    <Button
                        className={cn('button', { 'single-button': !showConnectButton })}
                        passiveColor={buttonPassiveColor}
                        hoverColor="green"
                        sizeAll="medium"
                        href={buyLink}
                        border={buttonBorder}
                        fontColor={buttonFontColor}
                        onClick={onClickBuy}
                    >
                        {buyButtonText}
                    </Button>
                }
                {showConnectButton &&
                    <TextLink
                        className={cn('detail-link')}
                        href={connectLink}
                        target="_blank"
                        onClick={onClickConnect}
                    >
                        {connectButtonText}
                    </TextLink>
                }
            </div>
        );
    }
}

export default ProductTileBuy;
