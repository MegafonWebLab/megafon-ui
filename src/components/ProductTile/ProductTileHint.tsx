import * as React from 'react';
import * as PropTypes from 'prop-types';
import { cnCreate } from '../../utils/cn';
import './style/ProductTileHint.less';
import { TLinkTargetType } from './ProductTile';

interface IProductTileHintProps {
    /** Title */
    title: string;
    /** Link href */
    linkHref?: string;
    /** Link target */
    linkTarget?: TLinkTargetType;
}

interface IElementAtributes {
    /** Href */
    href?: string;
    /** Target */
    target?: string;
}

const cn = cnCreate('mfui-product-tile-hint');
class ProductTileHint extends React.Component<IProductTileHintProps> {
    static propTypes = {
        linkHref: PropTypes.string,
        title: PropTypes.string,
        linkTarget: PropTypes.oneOf(['_self', '_blank', '_parent', '_top']),
    };

    static defaultProps = {
        linkTarget: '_self',
    };

    render() {
        const { title, linkHref, linkTarget } = this.props;
        const ElementType = linkHref ? 'a' : 'div';
        const attributes: IElementAtributes = linkHref ? { href: linkHref, target: linkTarget } : {};

        return (
            <div className={cn('')}>
                <ElementType {...attributes}>
                    <div className={cn('text')}>
                        <span className={cn('description')}>{title}</span>
                    </div>
                    <div className={cn('border')} />
                </ElementType>
            </div >
        );
    }
}

export default ProductTileHint;
