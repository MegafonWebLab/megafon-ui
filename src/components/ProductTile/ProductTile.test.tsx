import * as React from 'react';
import { shallow } from 'enzyme';
import ProductTile, {IProductTileProps , IServicePack } from './ProductTile';
import MegafonTv from 'icons/Service-logos/24/MF-TV.svg';

const tariff: Partial<IProductTileProps> = {
    buyButtonText: 'sdfsdf',
    connectButtonText: 'sdfsdf',
    secondParamsHead: 'sdfsdf',
    showConnectButton: true,
    title: 'Включайся! Смотри',
    'topBadgeTitle': 'sdfsdf',
    'link': '/tariffs/vklyuchaysya/smotri.html',
    'buyLink': '/zakaz/?tariff=look',
    'payment': {
        'value': '1000',
        'unitExtra': 'за 30 дней',
        'unitValue': '₽',
        'discount': '0',
    },
    'packs': [
        {
            'value': 1500,
            'unit': 'минут',
            'title': 'title',
            'isDelim': true,
        },
        {
            'value': 100,
            'unit': 'смс',
            'title': 'title',
            'isDelim': true,
        },
        {
            'value': 3,
            'unit': 'ГБ',
            'title': 'title',
            'isDelim': false,
        },
    ],
    'firstParams': {
        'items': [
            {
                'title': 'VIP-тариф',
                'caption': 'Привилегии обслуживания и другое',
                'svgIcon': <MegafonTv/>,
            },
            {
                'title': 'МегаФон ТВ',
                'caption': 'Безлимитный трафик и пакеты кино, сериалов и ТВ-каналов на любой вкус',
                'svgIcon': <MegafonTv/>,
            },
        ],
    },
    'secondParams': [
        {
            'title': 'Кэшбэк 20%',
            'footnote': 'Участвуйте в программе и тратьте кэшбэк',
            'value': '0',
            'unit': '',
            'svgIcon': <MegafonTv/>,
        },
        {
            'title': 'Безлимитные SMS',
            'footnote': '',
            'value': '55',
            'unit': '₽ за 30 дней',
            'svgIcon': <MegafonTv/>,
        },
    ],
    'cashback': {
        'title': 'Кэшбэк',
        'value': 20,
        'unit': '%',
    },
    'info': {
        'billingIds': [],
        'sms': '05007897',
        'smsText': 'ДА',
        'smsSite': null,
        'ivr': '05007897',
        'ussd': '<span class=\'c-number-call\'>sdfsdf</span>',
        'lkLink': 'https://lk.megafon.ru/tariffs/',
        'title': 'Включайся! Смотри',
        'id': 1598,
    },
};

const servicePacks: Array<Partial<IServicePack>> = [
    {
        'buyLink': '&options=null',
        'shopTag': 'tag1',
        'calls': {
            'value': 300,
            'unit': 'минут',
        },
        'traffic': {
            'value': 6,
            'unit': 'ГБ',
        },
        'payment': {
            'value': '400',
            'discount': '200',
        },
        'options': [
            {
                'title': 'Безлимитные SMS',
                'value': '55',
                'unit': '₽ за 30 дней',
                'footnote': '',
                'svgIcon': <MegafonTv />,
            },
            {
                'title': 'Кэшбэк 20%',
                'value': '0',
                'unit': '',
                'footnote': 'Участвуйте в программе и трпки в салонах и интернет-магазине МегаФона',
                'svgIcon': <MegafonTv />,
            },
        ],
    },
    {
        'buyLink': '&options=null',
        'shopTag': 'tag2',
        'calls': {
            'value': 650,
            'unit': 'минут',
        },
        'traffic': {
            'value': 6,
            'unit': 'ГБ',
        },
        'payment': {
            'value': '590',
            'discount': '300',
        },
        'options': [
            {
                'title': 'Безлимитные SMS',
                'value': '55',
                'unit': '₽ за 30 дней',
                'footnote': '',
                'svgIcon': <MegafonTv />,
            },
            {
                'title': 'Кэшбэк 20%',
                'value': '0',
                'unit': '',
                'footnote': 'Участвуйте тратьте кэшбэк на сервисы и покупки в салонах и магазине МегаФона',
                'svgIcon': <MegafonTv />,
            },
        ],
    },
    {
        'buyLink': '&options=null',
        'shopTag': 'tag3',
        'calls': {
            'value': 300,
            'unit': 'минут',
        },
        'traffic': {
            'value': 18,
            'unit': 'ГБ',
        },
        'payment': {
            'value': '590',
            'discount': '',
        },
        'options': [
            {
                'title': 'Безлимитные SMS',
                'value': '55',
                'unit': '₽ за 30 дней',
                'footnote': '',
                'svgIcon': <MegafonTv />,
            },
            {
                'title': 'Кэшбэк 20%',
                'value': '0',
                'unit': '',
                'footnote': 'Участвуйте в купки в салонах и интернет-магазине МегаФона',
                'svgIcon': <MegafonTv />,
            },
        ],
    },
    {
        'buyLink': '&options=null',
        'calls': {
            'value': 750,
            'unit': 'минут',
        },
        'traffic': {
            'value': 6,
            'unit': 'ГБ',
        },
        'payment': {
            'value': '670',
            'discount': '',
        },
        'options': [
            {
                'title': 'Безлимитные SMS',
                'value': '55',
                'unit': '₽ за 30 дней',
                'footnote': '',
                'svgIcon': <MegafonTv />,
            },
            {
                'title': 'Кэшбэк 20%',
                'value': '0',
                'unit': '',
                'footnote': 'Укупки в салонах и интернет-магазине МегаФона',
                'svgIcon': <MegafonTv />,
            },
        ],
    },
    {
        'buyLink': '&options=null',
        'calls': {
            'value': 900,
            'unit': 'минут',
        },
        'traffic': {
            'value': 6,
            'unit': 'ГБ',
        },
        'payment': {
            'value': '770',
            'discount': '',
        },
        'options': [
            {
                'title': 'Безлимитные SMS',
                'value': '55',
                'unit': '₽ за 30 дней',
                'footnote': '',
                'svgIcon': <MegafonTv />,
            },
            {
                'title': 'Кэшбэк 20%',
                'value': '0',
                'unit': '',
                'footnote': 'Участвуки в салонах и интернет-магазине МегаФона',
                'svgIcon': <MegafonTv />,
            },
        ],
    },
];

describe('<ProductTile />', () => {
    it('it renders ProductTile', () => {
        const wrapper = shallow(
            <ProductTile
                {...tariff}
                servicePacks={servicePacks}
                startCallsIndex={2}
                startTrafficIndex={1}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });

    describe('should call pased callback when click on range buttons', () => {
        it('should call onCallsChange when handleChangeCalls fired', () => {
            const onCallsChange = jest.fn();
            const wrapper = shallow<ProductTile>(
                <ProductTile
                    {...tariff}
                    servicePacks={servicePacks}
                    startCallsIndex={2}
                    startTrafficIndex={1}
                    onCallsChange={onCallsChange}
                />
            );

            wrapper.instance().handleChangeCalls({ target: {} } as React.SyntheticEvent<EventTarget>, '300' );
            expect(onCallsChange).toBeCalled();
        });

        it('should call onCallsChange when handleChangeCalls fired', () => {
            const onTrafficChange = jest.fn();
            const wrapper = shallow<ProductTile>(
                <ProductTile
                    {...tariff}
                    servicePacks={servicePacks}
                    startCallsIndex={2}
                    startTrafficIndex={1}
                    onTrafficChange={onTrafficChange}
                />
            );

            wrapper.instance().handleChangeTraffic({ target: {} } as React.SyntheticEvent<EventTarget>, '6' );
            expect(onTrafficChange).toBeCalled();
        });
    });
});
