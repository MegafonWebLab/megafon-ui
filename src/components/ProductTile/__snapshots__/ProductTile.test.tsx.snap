// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductTile /> it renders ProductTile 1`] = `
<div
  className="mfui-product-tile mfui-product-tile_constructor"
>
  <ProductTileHint
    title="sdfsdf"
  />
  <div
    className="mfui-product-tile__content"
  >
    <h2
      className="mfui-product-tile__title"
    >
      <TextLink
        className="mfui-product-tile__title-link"
        color="black"
        href="/tariffs/vklyuchaysya/smotri.html"
        onClick={[Function]}
        target="_blank"
        underlineStyle="none"
        underlineVisibility="hover"
      >
        Включайся! Смотри
      </TextLink>
    </h2>
    <div
      className="mfui-product-tile__link-wrapper"
    >
      <TextLink
        className="mfui-product-tile__detail-link"
        color="blue"
        href="/tariffs/vklyuchaysya/smotri.html"
        onClick={[Function]}
        target="_blank"
        underlineStyle="solid"
        underlineVisibility="hover"
      >
        Подробнее
      </TextLink>
    </div>
    <ProductTileCashback
      title="Кэшбэк"
      unit="%"
      value={20}
    />
    <ProductTilePrice
      discount="0"
      unitExtra="за 30 дней"
      unitValue="₽"
      value="1000"
    />
    <div
      className="mfui-product-tile__constructor"
    >
      <ProductTileDynamic
        currentPack={Object {}}
        onChangeCalls={[Function]}
        onChangeTraffic={[Function]}
        startCallsIndex={2}
        startTrafficIndex={1}
        switcher={
          Object {
            "calls": Array [
              "300",
              "650",
              "750",
              "900",
            ],
            "traffic": Array [
              "6",
              "18",
            ],
          }
        }
      />
      <ProductTileOptions
        options={
          Array [
            Object {
              "caption": "Привилегии обслуживания и другое",
              "svgIcon": <test-file-stub />,
              "title": "VIP-тариф",
            },
            Object {
              "caption": "Безлимитный трафик и пакеты кино, сериалов и ТВ-каналов на любой вкус",
              "svgIcon": <test-file-stub />,
              "title": "МегаФон ТВ",
            },
          ]
        }
      />
    </div>
    <ProductTileOptions
      head="sdfsdf"
      onClickBubble={[Function]}
      options={
        Array [
          Object {
            "footnote": "Участвуйте в программе и тратьте кэшбэк",
            "svgIcon": <test-file-stub />,
            "title": "Кэшбэк 20%",
            "unit": "",
            "value": "0",
          },
          Object {
            "footnote": "",
            "svgIcon": <test-file-stub />,
            "title": "Безлимитные SMS",
            "unit": "₽ за 30 дней",
            "value": "55",
          },
        ]
      }
    />
  </div>
  <ProductTileBuy
    buyButtonText="sdfsdf"
    buyLink="/zakaz/?tariff=look"
    connectButtonText="sdfsdf"
    onClickBuy={[Function]}
    onClickConnect={[Function]}
    showBuyButton={true}
    showConnectButton={true}
  />
</div>
`;
